text/0000-address.md

```diff
@@ -0,0 +1,232 @@
+- Feature Name: address
+- Start Date: 2019-01-17
+- RFC PR: (leave this empty)
+- Authors:
+  - Nicolas Di Prima (nicolas.diprima@iohk.io)
+  - Vincent Hanquez (vincent.hanquez@iohk.io)
+
+# Summary
+[summary]: #summary
+
+A new address format for the cardano blockchain that exposes extendable
+properties as well as being as compact as possible.
+
+The new proposed address format exposes the following properties:
+
+1. compact on the blockchain;
+2. extendable within reason;
+3. human readable encoding
+  * that includes visual discriminant;
+  * with checksum to detect errors;
+
+# Motivation
+[motivation]: #motivation
+
+The current address model is wasting quite a lot of space. It is a large
+CBOR structure and does not distinguish between what is necessary for the blockchain
+and what is necessary for the human readability. This RFC fixes this by
+proposing a much simpler address format in the sense that it is extendable
+within reason.
+
+1. the _compact-ability_ of the address on the blockchain is a prime motivation
+   of this proposal. Because of the model of the blockchain, every byte ever
+   included in the ledger will be stored there forever. We need to ensure that
+   we store only what is necessary. On the previous address encoding, the cbor
+   model allowed for including an unbounded byte array in the address which has been
+   leverage to include random data;
+2. Another motivation to keep the address as short as possible is that it will
+   save on transaction fees too (the current fee algorithm being linear in number
+   of bytes);
+3. The new address model allows to up to 123 different types of addresses. The
+   current RFC only proposes 2 models: simple and grouped.
+4. we also propose a new human readable encoding. This will allow visual
+   discrimination for the user between the old address format and the new one.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+An `Address` is a mean to associate a token to a cryptographic key pair.
+In order for Bob to send a token to Alice, Alice needs to generate a key
+pair, a `PublicKey` and a `PrivateKey`. Alice generates an `Address` from
+this `PublicKey` and sends the `Address` to Bob and Bob adds on the Ledger
+that he has sent a token to the public key.
+
+To create a new address, we only need the following information:
+
+* `PublicKey`: this will be necessary to be able to utilise the token;
+* `Discriminant`: this will prevent misuse of addresses between testing
+  environment and production environment;
+
+This is the most simple address. It is called a **Single Address**. This is
+because it is not associated with any staking key.
+
+It is possible to create a **Grouped Address**, it groups addresses to a
+staking key. To create a **Grouped Address** you need the `PublicKey`,
+the `Discriminant` (just like for a **Single Address**) and another
+`PublicKey` that is associating this address to another hierarchy.
```

Comment:
I think we should expand on this, to get the connection to what's in the design document. There, we have four new types of addresses (base, pointer, enterprise, and reward account). This RFC introduces two new kinds of addresses, which can be used to represent three of the addresses in the design document:

- base addresses can be implemented in terms of grouped addresses
- pointer addresses are not covered here
- enterprise addresses can be implemented in terms of single addresses
- reward account addresses can be implemented in terms of grouped addresses

I suggest using the address kind to distinguish the two kinds of grouped addresses (base and reward account), and adding pointer addresses.

Reply:
we're not going to support pointer address. they are not useful for MVP, can be bolted afterwards without a problem from the format PoV, and are raising *important* security questions.

Reply:
So are you saying that with this proposed address format, where the additional staking key hash takes only 32 bytes, the savings of space by using pointer addresses is not enough to warrant the additional complexity?

Reply:
@kantp can you share a document where types of addresses are defined - as I am not aware of the recent development

Reply:
@akegalj They are defined in the [design document for delegation](https://github.com/input-output-hk/fm-ledger-rules/tree/master/docs/delegation_design_spec).